src/core/dense_set.cc

```diff
@@ -0,0 +1,558 @@
+    return false;
+  }
+
+  DCHECK_LT(bucket_id, entries_.size());
+
+  // Try insert into flat surface first. Also handle the grow case
+  // if utilization is too high.
+  for (unsigned j = 0; j < 2; ++j) {
+    ChainVectorIterator list = FindEmptyAround(bucket_id);
+    if (list != entries_.end()) {
+      obj_malloc_used_ += PushFront(list, ptr);
+      if (std::distance(entries_.begin(), list) != bucket_id) {
+        list->SetDisplaced(std::distance(entries_.begin() + bucket_id, list));
+      }
+      ++num_used_buckets_;
+      ++size_;
+      return true;
+    }
+
+    if (size_ < entries_.size()) {
+      break;
+    }
+
+    Grow();
+    bucket_id = BucketId(hc);
+  }
+
+  DCHECK(!entries_[bucket_id].IsEmpty());
+
+  /**
+   * Since the current entry is not empty, it is either a valid chain
+   * or there is a displaced node here. In the latter case it is best to
+   * move the displaced node to its correct bucket. However there could be
+   * a displaced node there and so forth. Keep a stack of nodes that are displaced
+   * until we find a displaced node which hashes to an empty bucket.
+   * Then unwind the stack of pending displacements until each node is in
+   * its correct bucket
+   */
+  std::stack<std::pair<DensePtr, ChainVectorIterator>> to_fix_displacements;
+  auto curr = entries_.begin() + bucket_id;
+  while (!curr->IsEmpty() && curr->IsDisplaced()) {
+    DCHECK(!curr->IsLink());
+    DensePtr unlinked;
+    Unlink(ChainBegin(curr), nullptr, &unlinked);
+    --num_used_buckets_;
+
+    // undo the direction shift in a displaced pointer
+    curr -= unlinked.GetDisplacedDirection();
+    to_fix_displacements.push({unlinked, curr});
+  }
+
+  while (!to_fix_displacements.empty()) {
+    auto [ptr, list] = to_fix_displacements.top();
+    to_fix_displacements.pop();
+
+    if (!list->IsEmpty()) {
+      ++num_chain_entries_;
+    } else {
+      ++num_used_buckets_;
+    }
+
+    PushFront(list, &ptr);
+  }
+  if (!entries_[bucket_id].IsEmpty()) {
+    ++num_chain_entries_;
+  }
+
+  ChainVectorIterator list = entries_.begin() + bucket_id;
+  obj_malloc_used_ += PushFront(list, ptr);
+  DCHECK(!entries_[bucket_id].IsDisplaced());
+
+  ++size_;
+  return true;
+}
+
+std::pair<DenseSet::ChainIterator, DenseSet::ChainIterator> DenseSet::FindAround(void* ptr,
+                                                                                   uint32_t bid) {
+  // first look for displaced nodes since this is quicker than iterating a poential long chain
+  if (bid && Equal(&entries_[bid - 1], ptr)) {
+    return {ChainBegin(bid - 1), ChainEnd(bid - 1)};
+  }
+
+  if (bid + 1 < entries_.size() && Equal(&entries_[bid + 1], ptr)) {
+    return {ChainBegin(bid + 1), ChainEnd(bid + 1)};
+  }
+
+  // if the node is not displaced, search the correct chain
+  ChainIterator prev;
+  for (auto it = ChainBegin(bid); it != ChainEnd(bid); ++it) {
+    if (Equal(*it, ptr)) {
+      return {it, prev};
```

Comment:
just a suggestion: you need `prev` when erasing the item, right?
so you return in fact iterator to DenseLinkKey and the current. And then you can do `prev->next = whatever`.
prev->next points to current `it`. Why not return instead just 'DensePtr*` that is the reference to prev->next?
i.e. `DensePtr* FindAround(..)` will return `&entries_[..]`  if it's in the root of the chain, or &prev->next here.
You can still update your "next" by doing `*res = whatever` in the caller code. If that works - it's simpler and clearer.

Reply:
so originally I did want to use the double pointer approach to not need a previous pointer but because of the optimization of always trying to pop off a `DenseLinkKey` if possible, the objects between the current node and previous may need to be swapped. Again premature optimization and I ended up using this double pointer approach with minimal performance change.